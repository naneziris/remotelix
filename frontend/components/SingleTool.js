import React, { Component } from 'react';
import gql from 'graphql-tag';
import { Query } from 'react-apollo';
import Error from './ErrorMessage';
import styled from 'styled-components';
import Head from 'next/head';

const SingleToolStyles = styled.div`
  max-width: 1200px;
  margin: 2rem auto;
  box-shadow: ${props => props.theme.bs};
  display: grid;
  grid-auto-columns: 1fr;
  grid-auto-flow: column;
  min-height: 800px;
  img {
    width: 100%;
    height: 100%;
    object-fit: contain;
  }
  .details {
    margin: 3rem;
    font-size: 2rem;
  }
`;

const SINGLE_TOOL_QUERY = gql`
  query SINGLE_TOOL_QUERY($id: ID!) {
    tool(where: { id: $id }) {
      id
      title
      description
      largeImage
      url
      category
    }
  }
`;
class SingleTool extends Component {
  render() {
    return (
      <Query
        query={SINGLE_TOOL_QUERY}
        variables={{
          id: this.props.id,
        }}
      >
        {({ error, loading, data }) => {
          if (error) return <Error error={error} />;
          if (loading) return <p>Loading...</p>;
          if (!data.tool) return <p>No Tool Found for {this.props.id}</p>;
          const tool = data.tool;
          return (
            <SingleToolStyles>
              <Head>
                <title>Remotelix | {tool.title}</title>
              </Head>
              <img src={tool.largeImage} alt={tool.title} />
              <div className="details">
                <h2>Viewing {tool.title}</h2>
                <p>{tool.category}</p>
                <p>{tool.description}</p>
                <p>{tool.url}</p>
              </div>
            </SingleToolStyles>
          );
        }}
      </Query>
    );
  }
}

export default SingleTool;
export { SINGLE_TOOL_QUERY };
