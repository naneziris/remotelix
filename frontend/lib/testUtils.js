import casual from 'casual';

// seed it so we get consistent results
casual.seed(777);

const fakeTool = () => ({
  __typename: 'Tool',
  id: 'abc123',
  url: 'www.remotelix.com/tool/nikos',
  user: null,
  image: 'dog-small.jpg',
  title: 'dogs are best',
  titleToLowerCase: 'dogs are best',
  description: 'dogs',
  largeImage: 'dog.jpg',
  category: '',
});

const fakeUser = () => ({
  __typename: 'User',
  id: '4234',
  name: casual.name,
  email: casual.email,
  permissions: ['ADMIN'],
});

// Fake LocalStorage
class LocalStorageMock {
  constructor() {
    this.store = {};
  }

  clear() {
    this.store = {};
  }

  getItem(key) {
    return this.store[key] || null;
  }

  setItem(key, value) {
    this.store[key] = value.toString();
  }

  removeItem(key) {
    delete this.store[key];
  }
}

export {
  LocalStorageMock,
  fakeTool,
  fakeUser,
};
