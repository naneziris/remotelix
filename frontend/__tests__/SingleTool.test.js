import { mount } from 'enzyme';
import toJSON from 'enzyme-to-json';
import wait from 'waait';
import SingleTool, { SINGLE_TOOL_QUERY } from '../components/SingleTool';
import { MockedProvider } from 'react-apollo/test-utils';
import { fakeTool } from '../lib/testUtils';

describe('<SingleTool/>', () => {
  it('renders with proper data', async () => {
    const mocks = [
      {
        // when someone makes a request with this query and variable combo
        request: { query: SINGLE_TOOL_QUERY, variables: { id: '123' } },
        // return this fake data (mocked data)
        result: {
          data: {
            tool: fakeTool(),
          },
        },
      },
    ];
    const wrapper = mount(
      <MockedProvider mocks={mocks}>
        <SingleTool id="123" />
      </MockedProvider>
    );
    expect(wrapper.text()).toContain('Loading...');
    await wait();
    wrapper.update();
    // console.log(wrapper.debug());
    expect(toJSON(wrapper.find('h2'))).toMatchSnapshot();
    expect(toJSON(wrapper.find('img'))).toMatchSnapshot();
    expect(toJSON(wrapper.find('p'))).toMatchSnapshot();
  });

  it('Errors with a not found tool', async () => {
    const mocks = [
      {
        request: { query: SINGLE_TOOL_QUERY, variables: { id: '123' } },
        result: {
          errors: [{ message: 'Tools Not Found!' }],
        },
      },
    ];
    const wrapper = mount(
      <MockedProvider mocks={mocks}>
        <SingleTool id="123" />
      </MockedProvider>
    );
    await wait();
    wrapper.update();
    console.log(wrapper.debug());
    const tool = wrapper.find('[data-test="graphql-error"]');
    expect(tool.text()).toContain('Tools Not Found!');
    expect(toJSON(tool)).toMatchSnapshot();
  });
});
