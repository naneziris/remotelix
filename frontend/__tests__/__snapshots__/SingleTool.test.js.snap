// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SingleTool/> Errors with a not found tool 1`] = `
<p
  data-test="graphql-error"
>
  <strong>
    Oups!
  </strong>
  Tools Not Found!
</p>
`;

exports[`<SingleTool/> renders with proper data 1`] = `
<h2>
  Viewing 
  dogs are best
</h2>
`;

exports[`<SingleTool/> renders with proper data 2`] = `
<img
  alt="dogs are best"
  src="dog.jpg"
/>
`;

exports[`<SingleTool/> renders with proper data 3`] = `
Array [
  <p />,
  <p>
    dogs
  </p>,
  <p>
    www.remotelix.com/tool/nikos
  </p>,
]
`;
